verify.go

```diff
@@ -39,3 +43,86 @@ func VerifyInTotoEnvelope(ctx context.Context, env *dsse.Envelope, provider clie
 
 	return stmt, nil
 }
+
+func VerifyInTotoEnvelopeExt(ctx context.Context, env *Envelope, provider client.OpenIdProvider) (*intoto.Statement, error) {
+	tl, ok := ctx.Value(tlCtxKey(DefaultCtxKey)).(TL)
+	if !ok {
+		tl = &RekorTL{}
+	}
+
+	// enforce payload type
+	if env.PayloadType != intoto.PayloadType {
+		return nil, fmt.Errorf("unsupported payload type %s", env.PayloadType)
+	}
+
+	// verify signatures and transparency log entry
+	verifier := NewOPKSignerVerifier(provider)
+	for _, sig := range env.Signatures {
+		if sig.Extension.Kind != OpkSignatureID {
+			return nil, fmt.Errorf("error unsupported signature kind: %s", sig.Extension.Kind)
+		}
+
+		// verify opk signature
+		payload, err := base64.RawStdEncoding.DecodeString(env.Payload)
+		if err != nil {
+			return nil, fmt.Errorf("error failed to decode OPK payload: %w", err)
+		}
+		encPayload := dsse.PAE(intoto.PayloadType, payload)
+		pkToken := sig.Extension.Ext["pkt"].([]byte)
+		err = verifier.Verify(ctx, encPayload, pkToken)
+		if err != nil {
+			return nil, fmt.Errorf("error failed to verify PK token: %w", err)
+		}
+
+		// verify payload ephemeral ecdsa signature
+		ok, err := VerifyPayloadSignature(ctx, pkToken, encPayload, sig.Sig)
+		if !ok {
+			return nil, fmt.Errorf("error failed to verify payload signature: %w", err)
+		}
+
+		// verify TL entry
+		err = tl.VerifyLogEntry(ctx, sig.Extension.Ext["tl"].([]byte))
+		if err != nil {
+			return nil, fmt.Errorf("TL entry failed verification: %w", err)
+		}
+	}
+
+	// decode in-toto statement
+	stmt := new(intoto.Statement)
+	stmtBytes, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode in-toto statement: %w", err)
+	}
+	err = json.Unmarshal(stmtBytes, stmt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal in-toto statement: %w", err)
+	}
+	return stmt, nil
+}
+
+// VerifyPayloadSignature extracts the ephemeral ecdsa public key from a PK token and verifies the provided signature
+func VerifyPayloadSignature(ctx context.Context, pkToken []byte, payload []byte, signature string) (bool, error) {
+	token := &pktoken.PKToken{}
+	err := json.Unmarshal(pkToken, token)
+	if err != nil {
+		return false, fmt.Errorf("error unmarshalling PK token from JSON: %w", err)
+	}
+	cic, err := token.Cic.ProtectedHeaders().AsMap(ctx)
+	if err != nil {
+		return false, fmt.Errorf("error getting CIC protected headers: %w", err)
+	}
+	cicClaims, err := clientinstance.ParseClaims(cic)
+	if err != nil {
+		return false, fmt.Errorf("error failed to parse cic: %w", err)
+	}
+	decodedSig, err := base64.StdEncoding.DecodeString(signature)
+	if err != nil {
+		return false, fmt.Errorf("error failed to decode signature: %w", err)
+	}
+	ecPub := new(ecdsa.PublicKey)
+	err = cicClaims.PublicKey().Raw(ecPub)
+	if err != nil {
+		return false, fmt.Errorf("error failed to get public key from cic: %w", err)
+	}
+	return ecdsa.VerifyASN1(ecPub, s256(payload), decodedSig), nil
+}
```

Comment:
Tests?

Reply:
It is covered in: https://github.com/openpubkey/signed-attestation/blob/f3f8f32a8698557d532e039b759ad23f9668356d/sign_test.go#L70 but I will go throughout and see if I can add more tests.

Reply:
added [f8c9b55](https://github.com/openpubkey/signed-attestation/pull/4/commits/f8c9b55e7dbd4fc2aad10da461a5f00c3ee08a04)